internal/worker/providertracker/manifold.go

```diff
@@ -152,3 +169,97 @@ func (f serviceFactory) Config() ConfigService {
 func (f serviceFactory) Credential() CredentialService {
 	return f.factory.Credential()
 }
+
+// IAASGetProvider creates a new provider from the given args.
+func IAASGetProvider(newProvider ProviderFunc[environs.Environ]) func(ctx context.Context, getter ProviderConfigGetter) (environs.Environ, environscloudspec.CloudSpec, error) {
+	return func(ctx context.Context, getter ProviderConfigGetter) (environs.Environ, environscloudspec.CloudSpec, error) {
+		// We can't use newProvider directly, as type invariance prevents us
+		// from using it with the environs.GetEnvironAndCloud function.
+		// Just wrap it in a closure to work around this.
+		provider, spec, err := environs.GetEnvironAndCloud(ctx, getter, func(ctx context.Context, op environs.OpenParams) (environs.Environ, error) {
+			return newProvider(ctx, op)
+		})
+		if err != nil {
+			return nil, environscloudspec.CloudSpec{}, errors.Trace(err)
+		}
+		return provider, *spec, nil
+	}
+}
+
+// CAASGetProvider creates a new provider from the given args.
+func CAASGetProvider(newProvider ProviderFunc[caas.Broker]) func(ctx context.Context, getter ProviderConfigGetter) (caas.Broker, environscloudspec.CloudSpec, error) {
+	return func(ctx context.Context, getter ProviderConfigGetter) (caas.Broker, environscloudspec.CloudSpec, error) {
+		cloudSpec, err := getter.CloudSpec(ctx)
+		if err != nil {
+			return nil, environscloudspec.CloudSpec{}, errors.Annotate(err, "cannot get cloud information")
+		}
+
+		cfg, err := getter.ModelConfig(ctx)
+		if err != nil {
+			return nil, environscloudspec.CloudSpec{}, errors.Trace(err)
+		}
+
+		broker, err := newProvider(ctx, environs.OpenParams{
+			ControllerUUID: getter.ControllerUUID().String(),
+			Cloud:          cloudSpec,
+			Config:         cfg,
+		})
+		if err != nil {
+			return nil, environscloudspec.CloudSpec{}, errors.Annotate(err, "cannot create caas broker")
+		}
+		return broker, cloudSpec, nil
+	}
+}
+
+func manifoldOutput[T Provider](in worker.Worker, out any) error {
+	// In order to switch on the type of the provider, we need to use a type
+	// assertion to get the underlying value.
+	switch any(new(T)).(type) {
```

Comment:
For those playing along at home, we discussed this at length and tried various alternative approaches, none of which made it any nicer than this brutish type switching.

Reply:
An alternative is to use `var t T` which will give you a zero type and then we can use reflection to switch on the type. Neither of these are palatable.